File: jam-pvm/src/main/kotlin/io/forge/jam/pvm/engine/BasicMemory.kt
package io.forge.jam.pvm.engine

import io.forge.jam.pvm.PvmConstants

/**
 * Basic memory implementation for non-dynamic paging mode.
 * Implements the Memory interface with per-page access control via PageMap.
 */
class BasicMemory private constructor(
    private var _pageMap: PageMap,
    private val rwData: MutableList<UByte>,
    private val stack: MutableList<UByte>,
    private val aux: MutableList<UByte>,
    private var isMemoryDirty: Boolean,
    private var _heapSize: UInt
) : Memory {

    override val pageMap: PageMap get() = _pageMap

    companion object {
        fun new(): BasicMemory = BasicMemory(
            _pageMap = PageMap(PvmConstants.ZP),
            rwData = mutableListOf(),
            stack = mutableListOf(),
            aux = mutableListOf(),
            isMemoryDirty = false,
            _heapSize = 0u
        )

        fun MutableList<UByte>.resize(newSize: Int, padValue: UByte = 0u) {
            when {
                size > newSize -> subList(newSize, size).clear()
                size < newSize -> addAll(List(newSize - size) { padValue })
            }
        }

        fun alignToNextPageSize(pageSize: Int, size: Int): Int {
            return ((size + pageSize - 1) / pageSize) * pageSize
        }
    }

    /**
     * Returns the current heap size.
     */
    override fun heapSize(): UInt = _heapSize

    /**
     * Checks if a page is mapped in the current memory layout.
     */
    fun isPageMapped(module: Module, address: UInt): Boolean {
        val memoryMap = module.memoryMap()
        return when {
            address >= memoryMap.auxDataAddress ->
                address < memoryMap.auxDataAddress + memoryMap.auxDataSize

            address >= memoryMap.stackAddressLow() ->
                address < memoryMap.stackAddressLow() + memoryMap.stackSize

            address >= memoryMap.rwDataAddress -> {
                val heapEnd = memoryMap.heapBase + _heapSize
                val effectiveEnd = maxOf(memoryMap.rwDataAddress + memoryMap.rwDataSize, heapEnd)
                address < effectiveEnd
            }

            address >= memoryMap.roDataAddress() ->
                module.interpretedModule() != null &&
                    address < memoryMap.roDataAddress() + module.interpretedModule()!!.roData.size.toUInt()

            else -> false
        }
    }

    /**
     * Marks the memory as dirty, requiring a reset before next use.
     */
    fun markDirty() {
        isMemoryDirty = true
    }

    /**
     * Resets the memory if it's marked as dirty.
     */
    fun reset(module: Module) {
        if (isMemoryDirty) {
            forceReset(module)
        }
    }

    /**
     * Forces a reset of all memory regions and reinitializes PageMap.
     */
    fun forceReset(module: Module) {
        rwData.clear()
        stack.clear()
        aux.clear()
        isMemoryDirty = false

        module.interpretedModule()?.let { interpretedModule ->
            val memoryMap = module.memoryMap()

            // Extend RW data from interpreted module
            rwData.addAll(interpretedModule.rwData.map { it.toUByte() })

            // Resize memory regions to match memory map
            rwData.resize(memoryMap.rwDataSize.toInt())
            stack.resize(memoryMap.stackSize.toInt())
            aux.resize(memoryMap.auxDataSize.toInt())

            // The interpretedModule.heapEmptyPages contains the original heapPages from blob.
            val actualRwDataLen = (interpretedModule.rwData.size).toUInt()
            val pageAlignedRwDataLen = alignToNextPageSize(memoryMap.pageSize.toInt(), actualRwDataLen.toInt()).toUInt()
            val heapEmptyPagesSize = interpretedModule.heapEmptyPages.toUInt() * memoryMap.pageSize
            _heapSize = memoryMap.rwDataAddress + pageAlignedRwDataLen + heapEmptyPagesSize - memoryMap.heapBase
            println("[HEAP-INIT] actualRwDataLen=$actualRwDataLen, pageAlignedRwDataLen=$pageAlignedRwDataLen, heapEmptyPagesSize=$heapEmptyPagesSize, heapPages=${interpretedModule.heapEmptyPages}, _heapSize=$_heapSize")

            initializePageMap(module, memoryMap)
        }
    }

    /**
     * Initializes the PageMap with correct page access permissions.
     */
    private fun initializePageMap(module: Module, memoryMap: io.forge.jam.pvm.Abi.MemoryMap) {
        val pageMapEntries = mutableListOf<Triple<UInt, UInt, PageAccess>>()

        // RO data region - READ_ONLY
        module.interpretedModule()?.let { interpretedModule ->
            if (interpretedModule.roData.isNotEmpty()) {
                pageMapEntries.add(
                    Triple(
                        memoryMap.roDataAddress(),
                        memoryMap.roDataSize,
                        PageAccess.READ_ONLY
                    )
                )
            }
        }

        // RW data region (including heap) - READ_WRITE
        if (memoryMap.rwDataSize > 0u) {
            pageMapEntries.add(
                Triple(
                    memoryMap.rwDataAddress,
                    memoryMap.rwDataSize,
                    PageAccess.READ_WRITE
                )
            )
        }

        // Stack region - READ_WRITE
        if (memoryMap.stackSize > 0u) {
            pageMapEntries.add(
                Triple(
                    memoryMap.stackAddressLow(),
                    memoryMap.stackSize,
                    PageAccess.READ_WRITE
                )
            )
        }

        // Aux data region - typically READ_ONLY, but GP stack region is READ_WRITE
        if (memoryMap.auxDataSize > 0u) {
            // Check if GP stack region overlaps with aux data
            val gpStackLow = PvmConstants.GP_STACK_LOW
            val gpStackBase = PvmConstants.GP_STACK_BASE

            if (gpStackLow >= memoryMap.auxDataAddress &&
                gpStackBase <= memoryMap.auxDataAddress + memoryMap.auxDataSize
            ) {
                // GP stack is within aux data - mark it as READ_WRITE
                // Mark everything before GP stack as READ_ONLY
                if (gpStackLow > memoryMap.auxDataAddress) {
                    pageMapEntries.add(
                        Triple(
                            memoryMap.auxDataAddress,
                            gpStackLow - memoryMap.auxDataAddress,
                            PageAccess.READ_ONLY
                        )
                    )
                }

                // GP stack region - READ_WRITE
                pageMapEntries.add(
                    Triple(
                        gpStackLow,
                        PvmConstants.GP_STACK_SIZE,
                        PageAccess.READ_WRITE
                    )
                )

                // Mark everything after GP stack as READ_ONLY
                val afterGpStack = gpStackBase
                val auxEnd = memoryMap.auxDataAddress + memoryMap.auxDataSize
                if (afterGpStack < auxEnd) {
                    pageMapEntries.add(
                        Triple(
                            afterGpStack,
                            auxEnd - afterGpStack,
                            PageAccess.READ_ONLY
                        )
                    )
                }
            } else {
                // No GP stack overlap - mark entire aux as READ_ONLY
                pageMapEntries.add(
                    Triple(
                        memoryMap.auxDataAddress,
                        memoryMap.auxDataSize,
                        PageAccess.READ_ONLY
                    )
                )
            }
        }

        _pageMap = PageMap.create(pageMapEntries, memoryMap.pageSize)
    }

    // ========== Memory Interface Implementation ==========

    override fun read(address: UInt): UByte {
        ensureReadable(address, 1)
        val bytes = readInternal(address, 1u)
            ?: throw MemoryError.NotReadable(address)
        return bytes[0].toUByte()
    }

    override fun read(address: UInt, length: Int): ByteArray {
        if (length == 0) return ByteArray(0)
        ensureReadable(address, length)
        return readInternal(address, length.toUInt())
            ?: throw MemoryError.NotReadable(address)
    }

    override fun write(address: UInt, value: UByte) {
        ensureWritable(address, 1)
        writeInternal(address, byteArrayOf(value.toByte()))
    }

    override fun write(address: UInt, values: ByteArray) {
        if (values.isEmpty()) return
        ensureWritable(address, values.size)
        writeInternal(address, values)
    }

    override fun sbrk(size: UInt): UInt {
        throw UnsupportedOperationException("sbrk requires Module context, use sbrk(module, size)")
    }

    /**
     * Implements the sbrk system call to grow heap memory.
     * Returns the PREVIOUS heap end (base of newly allocated memory).
     */
    fun sbrk(module: Module, size: UInt): UInt? {
        val memoryMap = module.memoryMap()

        // Calculate current heap end (heapBase + heapSize)
        val prevHeapEnd = memoryMap.heapBase + _heapSize

        println(
            "[SBRK-KOTLIN] size=$size, heapBase=${memoryMap.heapBase}, _heapSize=$_heapSize, prevHeapEnd=$prevHeapEnd, instanceId=${
                System.identityHashCode(
                    this
                )
            }"
        )
        println("[SBRK-KOTLIN] rwDataAddress=${memoryMap.rwDataAddress}, rwDataSize=${memoryMap.rwDataSize}, maxHeapSize=${memoryMap.maxHeapSize}, rwData.size=${rwData.size}")

        // If size is 0, just return current heap end
        if (size == 0u) {
            return prevHeapEnd
        }

        // Check for overflow
        val newHeapSize = _heapSize.plus(size).takeIf { it >= _heapSize } ?: return null

        if (newHeapSize > memoryMap.maxHeapSize) {
            println("[SBRK-KOTLIN] FAILED: newHeapSize=$newHeapSize > maxHeapSize=${memoryMap.maxHeapSize}")
            return null
        }
        println("[SBRK-KOTLIN] OK: newHeapSize=$newHeapSize, newHeapEnd=${memoryMap.heapBase + newHeapSize}")

        _heapSize = newHeapSize
        val newHeapEnd = memoryMap.heapBase + newHeapSize

        // Expand rwData if needed
        if (newHeapEnd.toInt() > memoryMap.rwDataAddress.toInt() + rwData.size) {
            val newSize = alignToNextPageSize(
                pageSize = memoryMap.pageSize.toInt(),
                size = newHeapEnd.toInt()
            ) - memoryMap.rwDataAddress.toInt()

            println("[SBRK-KOTLIN] Expanding rwData: oldSize=${rwData.size}, newSize=$newSize, newHeapEnd=$newHeapEnd")
            rwData.resize(newSize)
        }

        // Update PageMap for newly allocated pages
        val prevPageBoundary = alignToNextPageSize(memoryMap.pageSize.toInt(), prevHeapEnd.toInt())
        if (newHeapEnd.toInt() > prevPageBoundary) {
            val startPage = prevPageBoundary.toUInt() / memoryMap.pageSize
            val endPage =
                alignToNextPageSize(memoryMap.pageSize.toInt(), newHeapEnd.toInt()).toUInt() / memoryMap.pageSize
            val pageCount = (endPage - startPage).toInt()
            if (pageCount > 0) {
                println("[SBRK-KOTLIN] Adding pages: startPage=$startPage, endPage=$endPage, pageCount=$pageCount")
                _pageMap.updatePages(startPage, pageCount, PageAccess.READ_WRITE)
            }
        }

        // Return the PREVIOUS heap end (base of newly allocated memory)
        return prevHeapEnd
    }

    // ========== Internal Methods ==========

    /**
     * Gets a read-only slice of memory (internal implementation).
     */
    private fun readInternal(address: UInt, length: UInt): ByteArray? {
        val memoryMap = getMemoryMapFromContext() ?: return null

        val (start, memorySlice) = when {
            address >= memoryMap.auxDataAddress ->
                memoryMap.auxDataAddress to aux

            // GP standard stack region (readable from aux data space)
            address >= PvmConstants.GP_STACK_LOW && address < PvmConstants.GP_STACK_BASE ->
                memoryMap.auxDataAddress to aux

            address >= memoryMap.stackAddressLow() ->
                memoryMap.stackAddressLow() to stack

            address >= memoryMap.rwDataAddress -> {
                memoryMap.rwDataAddress to rwData
            }

            address >= memoryMap.roDataAddress() -> {
                val interpretedModule = currentModule?.interpretedModule() ?: return null
                memoryMap.roDataAddress() to interpretedModule.roData.map { it.toUByte() }
            }

            else -> return null
        }

        val offset = (address - start).toInt()
        return memorySlice
            .subList(offset, (offset + length.toInt()).coerceAtMost(memorySlice.size))
            .map { it.toByte() }
            .toByteArray()
    }

    /**
     * Writes to memory (internal implementation).
     */
    private fun writeInternal(address: UInt, data: ByteArray) {
        val memoryMap = getMemoryMapFromContext() ?: return

        val (start, memorySlice) = when {
            // GP standard stack region is writable (falls within aux data space)
            address >= PvmConstants.GP_STACK_LOW && address < PvmConstants.GP_STACK_BASE ->
                memoryMap.auxDataAddress to aux

            address >= memoryMap.stackAddressLow() ->
                memoryMap.stackAddressLow() to stack

            address >= memoryMap.rwDataAddress ->
                memoryMap.rwDataAddress to rwData

            else -> return
        }

        isMemoryDirty = true
        val offset = (address - start).toInt()
        data.forEachIndexed { index, byte ->
            if (offset + index < memorySlice.size) {
                memorySlice[offset + index] = byte.toUByte()
            }
        }
    }

    // Thread-local storage for current module context
    private var currentModule: Module? = null

    /**
     * Sets the current module context for memory operations.
     */
    fun setModuleContext(module: Module) {
        currentModule = module
    }

    private fun getMemoryMapFromContext(): io.forge.jam.pvm.Abi.MemoryMap? {
        return currentModule?.memoryMap()
    }

    // ========== Legacy Methods for Backward Compatibility ==========

    /**
     * Gets a read-only slice of memory (legacy method for Visitor compatibility).
     */
    fun getMemorySlice(module: Module, address: UInt, length: UInt): ByteArray? {
        currentModule = module
        return readInternal(address, length)
    }

    /**
     * Gets a mutable slice of memory (legacy method for Visitor compatibility).
     */
    fun getMemorySliceMut(
        module: Module,
        address: UInt,
        length: UInt,
        isExternal: Boolean = false
    ): MutableList<UByte>? {
        currentModule = module
        val memoryMap = module.memoryMap()

        val (start, memorySlice) = when {
            isExternal && address >= memoryMap.auxDataAddress ->
                memoryMap.auxDataAddress to aux

            // GP standard stack region is writable (falls within aux data space)
            address >= PvmConstants.GP_STACK_LOW && address < PvmConstants.GP_STACK_BASE ->
                memoryMap.auxDataAddress to aux

            address >= memoryMap.stackAddressLow() ->
                memoryMap.stackAddressLow() to stack

            address >= memoryMap.rwDataAddress ->
                memoryMap.rwDataAddress to rwData

            else -> return null
        }

        isMemoryDirty = true
        val offset = (address - start).toInt()
        return memorySlice.subList(offset, (offset + length.toInt()).coerceAtMost(memorySlice.size))
    }

    /**
     * Checks if a memory range is writable (legacy method for Visitor compatibility).
     */
    fun isWritable(module: Module, address: UInt, length: UInt): Boolean {
        // Use PageMap for access control
        return pageMap.isWritable(address, length.toInt()).first
    }
}

/**
 * Extension function for BasicMemory to check readability.
 */
private fun BasicMemory.ensureReadable(address: UInt, length: Int) {
    val readResult = pageMap.isReadable(address, length)
    if (!readResult.first) {
        throw MemoryError.NotReadable(readResult.second)
    }
}

/**
 * Extension function for BasicMemory to check writability.
 */
private fun BasicMemory.ensureWritable(address: UInt, length: Int) {
    val writeResult = pageMap.isWritable(address, length)
    if (!writeResult.first) {
        throw MemoryError.NotWritable(writeResult.second)
    }
}

File: modules/pvm/src/main/scala/io/forge/jam/pvm/memory/BasicMemory.scala
package io.forge.jam.pvm.memory

import spire.math.{UByte, UShort, UInt, ULong}
import io.forge.jam.pvm.{MemoryResult, MemoryMap, Abi, PvmConstants}
import io.forge.jam.pvm.types.*

/**
 * Basic memory implementation for non-dynamic paging mode.
 *
 * Implements the Memory trait with per-page access control via PageMap.
 * Manages separate memory regions: roData, rwData, stack, heap, and aux.
 */
final class BasicMemory private (
  private var _pageMap: PageMap,
  private val memoryMap: MemoryMap,
  private var roData: Array[Byte],
  private val rwData: Array[Byte],
  private val stack: Array[Byte],
  private val aux: Array[Byte],
  private var _heapSize: UInt,
  private var isDirty: Boolean
) extends Memory:

  override def pageMap: PageMap = _pageMap

  override def heapSize: UInt = _heapSize

  override def heapEnd: UInt = memoryMap.heapBase + _heapSize

  // ============================================================================
  // Load Operations
  // ============================================================================

  override def loadU8(address: UInt): MemoryResult[UByte] =
    checkReadable(address, 1) match
      case Some(err) => err
      case None =>
        readByte(address) match
          case Some(b) => MemoryResult.Success(UByte(b))
          case None => MemoryResult.OutOfBounds(address)

  override def loadU16(address: UInt): MemoryResult[UShort] =
    checkReadable(address, 2) match
      case Some(err) => err
      case None =>
        val b0 = readByte(address)
        val b1 = readByte(address + UInt(1))
        (b0, b1) match
          case (Some(lo), Some(hi)) =>
            val value = ((hi & 0xFF) << 8) | (lo & 0xFF)
            MemoryResult.Success(UShort(value.toShort))
          case _ => MemoryResult.OutOfBounds(address)

  override def loadU32(address: UInt): MemoryResult[UInt] =
    checkReadable(address, 4) match
      case Some(err) => err
      case None =>
        val bytes = readBytes(address, 4)
        bytes match
          case Some(arr) =>
            val value = (arr(0) & 0xFF) |
              ((arr(1) & 0xFF) << 8) |
              ((arr(2) & 0xFF) << 16) |
              ((arr(3) & 0xFF) << 24)
            MemoryResult.Success(UInt(value))
          case None => MemoryResult.OutOfBounds(address)

  override def loadU64(address: UInt): MemoryResult[ULong] =
    checkReadable(address, 8) match
      case Some(err) => err
      case None =>
        val bytes = readBytes(address, 8)
        bytes match
          case Some(arr) =>
            var value = 0L
            var i = 0
            while i < 8 do
              value |= (arr(i) & 0xFFL) << (i * 8)
              i += 1
            MemoryResult.Success(ULong(value))
          case None => MemoryResult.OutOfBounds(address)

  // ============================================================================
  // Store Operations
  // ============================================================================

  override def storeU8(address: UInt, value: UByte): MemoryResult[Unit] =
    checkWritable(address, 1) match
      case Some(err) => err
      case None =>
        if writeByte(address, value.toByte) then
          isDirty = true
          MemoryResult.Success(())
        else
          MemoryResult.OutOfBounds(address)

  override def storeU16(address: UInt, value: UShort): MemoryResult[Unit] =
    checkWritable(address, 2) match
      case Some(err) => err
      case None =>
        val ok1 = writeByte(address, (value.toInt & 0xFF).toByte)
        val ok2 = writeByte(address + UInt(1), ((value.toInt >> 8) & 0xFF).toByte)
        if ok1 && ok2 then
          isDirty = true
          MemoryResult.Success(())
        else
          MemoryResult.OutOfBounds(address)

  override def storeU32(address: UInt, value: UInt): MemoryResult[Unit] =
    checkWritable(address, 4) match
      case Some(err) => err
      case None =>
        val v = value.signed
        val bytes = Array(
          (v & 0xFF).toByte,
          ((v >> 8) & 0xFF).toByte,
          ((v >> 16) & 0xFF).toByte,
          ((v >> 24) & 0xFF).toByte
        )
        if writeBytes(address, bytes) then
          isDirty = true
          MemoryResult.Success(())
        else
          MemoryResult.OutOfBounds(address)

  override def storeU64(address: UInt, value: ULong): MemoryResult[Unit] =
    checkWritable(address, 8) match
      case Some(err) => err
      case None =>
        val v = value.signed
        val bytes = new Array[Byte](8)
        var i = 0
        while i < 8 do
          bytes(i) = ((v >> (i * 8)) & 0xFF).toByte
          i += 1
        if writeBytes(address, bytes) then
          isDirty = true
          MemoryResult.Success(())
        else
          MemoryResult.OutOfBounds(address)

  // ============================================================================
  // Bulk Operations
  // ============================================================================

  override def getMemorySlice(address: UInt, length: Int): MemoryResult[Array[Byte]] =
    if length == 0 then return MemoryResult.Success(Array.empty)

    checkReadable(address, length) match
      case Some(err) => err
      case None =>
        readBytes(address, length) match
          case Some(arr) => MemoryResult.Success(arr)
          case None => MemoryResult.OutOfBounds(address)

  override def setMemorySlice(address: UInt, data: Array[Byte]): MemoryResult[Unit] =
    if data.isEmpty then return MemoryResult.Success(())

    checkWritable(address, data.length) match
      case Some(err) => err
      case None =>
        if writeBytes(address, data) then
          isDirty = true
          MemoryResult.Success(())
        else
          MemoryResult.OutOfBounds(address)

  // ============================================================================
  // Heap Operations (sbrk)
  // ============================================================================

  override def sbrk(size: UInt): Option[UInt] =
    val prevHeapEnd = heapEnd

    // sbrk(0) just returns current heap end
    if size == UInt(0) then return Some(prevHeapEnd)

    // Check for overflow
    val newHeapSizeLong = _heapSize.toLong + size.toLong
    if newHeapSizeLong > 0xFFFFFFFFL then return None

    val newHeapSize = UInt(newHeapSizeLong.toInt)

    // Check against max heap size
    if newHeapSize.toLong > memoryMap.maxHeapSize.toLong then return None

    _heapSize = newHeapSize
    val newHeapEnd = memoryMap.heapBase + newHeapSize

    // Update page map for new heap pages
    val prevPageBoundary = alignToNextPageSize(memoryMap.pageSize, prevHeapEnd)
    if newHeapEnd.toLong > prevPageBoundary.toLong then
      val startPage = UInt((prevPageBoundary.toLong / memoryMap.pageSize.toLong).toInt)
      val endPage = UInt((alignToNextPageSize(memoryMap.pageSize, newHeapEnd).toLong / memoryMap.pageSize.toLong).toInt)
      val pageCount = (endPage.signed - startPage.signed)
      if pageCount > 0 then
        _pageMap.updatePages(startPage, pageCount, PageAccess.ReadWrite)

    Some(prevHeapEnd)

  // ============================================================================
  // Internal Helpers
  // ============================================================================

  private def checkReadable(address: UInt, length: Int): Option[MemoryResult[Nothing]] =
    val (isReadable, failAddr) = _pageMap.isReadable(address, length)
    if !isReadable then
      Some(MemoryResult.Segfault(address, _pageMap.alignToPageStart(failAddr)))
    else
      None

  private def checkWritable(address: UInt, length: Int): Option[MemoryResult[Nothing]] =
    val (isWritable, failAddr) = _pageMap.isWritable(address, length)
    if !isWritable then
      Some(MemoryResult.Segfault(address, _pageMap.alignToPageStart(failAddr)))
    else
      None

  /**
   * Reads a single byte from the appropriate region.
   */
  private def readByte(address: UInt): Option[Byte] =
    val addr = address.signed
    val addrLong = address.toLong

    // Check RO data region
    val roStart = memoryMap.roDataAddress.signed
    val roEnd = roStart + roData.length
    if addr >= roStart && addr < roEnd then
      return Some(roData(addr - roStart))
    else if addr >= roStart && addr < (roStart + memoryMap.roDataSize.signed) then
      // Within RO region but past actual data - return 0
      return Some(0.toByte)

    // Check RW data + heap region
    val rwStart = memoryMap.rwDataAddress.signed
    val heapEndAddr = (memoryMap.heapBase + _heapSize).signed
    val effectiveEnd = math.max(rwStart + memoryMap.rwDataSize.signed, heapEndAddr)
    if addr >= rwStart && addr < effectiveEnd then
      val offset = addr - rwStart
      if offset >= 0 && offset < rwData.length then
        return Some(rwData(offset))
      else
        return Some(0.toByte) // Beyond buffer but in valid heap region

    // Check stack region
    val stackLow = memoryMap.stackAddressLow.signed
    val stackHigh = memoryMap.stackAddressHigh.signed
    if addr >= stackLow && addr < stackHigh then
      val offset = addr - stackLow
      if offset >= 0 && offset < stack.length then
        return Some(stack(offset))
      else
        return Some(0.toByte)

    // Check aux data region
    val auxStart = memoryMap.auxDataAddress.signed
    val auxEnd = auxStart + memoryMap.auxDataSize.signed
    if addr >= auxStart && addr < auxEnd then
      val offset = addr - auxStart
      if offset >= 0 && offset < aux.length then
        return Some(aux(offset))
      else
        return Some(0.toByte)

    None

  /**
   * Reads multiple bytes from memory.
   */
  private def readBytes(address: UInt, length: Int): Option[Array[Byte]] =
    val result = new Array[Byte](length)
    var i = 0
    while i < length do
      readByte(address + UInt(i)) match
        case Some(b) => result(i) = b
        case None => return None
      i += 1
    Some(result)

  /**
   * Writes a single byte to the appropriate region.
   */
  private def writeByte(address: UInt, value: Byte): Boolean =
    val addr = address.signed

    // RO data region - not writable
    val roStart = memoryMap.roDataAddress.signed
    val roEnd = roStart + memoryMap.roDataSize.signed
    if addr >= roStart && addr < roEnd then
      return false

    // Check RW data + heap region
    val rwStart = memoryMap.rwDataAddress.signed
    val heapEndAddr = (memoryMap.heapBase + _heapSize).signed
    val effectiveEnd = math.max(rwStart + memoryMap.rwDataSize.signed, heapEndAddr)
    if addr >= rwStart && addr < effectiveEnd then
      val offset = addr - rwStart
      if offset >= 0 && offset < rwData.length then
        rwData(offset) = value
        return true
      else
        // Within valid heap but beyond buffer - would need lazy expansion
        return false

    // Check stack region
    val stackLow = memoryMap.stackAddressLow.signed
    val stackHigh = memoryMap.stackAddressHigh.signed
    if addr >= stackLow && addr < stackHigh then
      val offset = addr - stackLow
      if offset >= 0 && offset < stack.length then
        stack(offset) = value
        return true
      else
        return false

    // Check aux data region (GP stack portion is writable)
    val gpStackLow = PvmConstants.GpStackLow.signed
    val gpStackBase = PvmConstants.GpStackBase.signed
    if addr >= gpStackLow && addr < gpStackBase then
      val auxStart = memoryMap.auxDataAddress.signed
      val offset = addr - auxStart
      if offset >= 0 && offset < aux.length then
        aux(offset) = value
        return true

    false

  /**
   * Writes multiple bytes to memory.
   */
  private def writeBytes(address: UInt, data: Array[Byte]): Boolean =
    var i = 0
    while i < data.length do
      if !writeByte(address + UInt(i), data(i)) then
        return false
      i += 1
    true

  /**
   * Aligns a size up to the next page boundary.
   */
  private def alignToNextPageSize(pageSize: UInt, size: UInt): UInt =
    val ps = pageSize.toLong
    val sz = size.toLong
    UInt((((sz + ps - 1) / ps) * ps).toInt)

  /**
   * Marks memory as dirty (needs reset before reuse).
   */
  def markDirty(): Unit = isDirty = true

  /**
   * Returns true if memory has been modified.
   */
  def dirty: Boolean = isDirty

  /**
   * Sets the read-only data (typically from program blob).
   */
  def setRoData(data: Array[Byte]): Unit =
    roData = data

object BasicMemory:
  /**
   * Creates a new BasicMemory instance from a MemoryMap.
   */
  def create(memoryMap: MemoryMap): BasicMemory =
    val pageMap = initializePageMap(memoryMap, None)

    new BasicMemory(
      _pageMap = pageMap,
      memoryMap = memoryMap,
      roData = new Array[Byte](memoryMap.roDataSize.signed),
      rwData = new Array[Byte](memoryMap.rwDataSize.signed),
      stack = new Array[Byte](memoryMap.stackSize.signed),
      aux = new Array[Byte](memoryMap.auxDataSize.signed),
      _heapSize = UInt(0),
      isDirty = false
    )

  /**
   * Creates a BasicMemory with initial RO/RW data.
   */
  def create(memoryMap: MemoryMap, roData: Array[Byte], rwData: Array[Byte]): BasicMemory =
    val pageMap = initializePageMap(memoryMap, Some(roData.length))

    // Copy rwData into buffer
    val rwBuffer = new Array[Byte](memoryMap.rwDataSize.signed)
    System.arraycopy(rwData, 0, rwBuffer, 0, math.min(rwData.length, rwBuffer.length))

    new BasicMemory(
      _pageMap = pageMap,
      memoryMap = memoryMap,
      roData = roData,
      rwData = rwBuffer,
      stack = new Array[Byte](memoryMap.stackSize.signed),
      aux = new Array[Byte](memoryMap.auxDataSize.signed),
      _heapSize = UInt(0),
      isDirty = false
    )

  /**
   * Initializes the PageMap with correct page access permissions.
   */
  private def initializePageMap(memoryMap: MemoryMap, roDataActualSize: Option[Int]): PageMap =
    val mappings = scala.collection.mutable.ListBuffer[(UInt, UInt, PageAccess)]()

    // RO data region - READ_ONLY
    val roSize = roDataActualSize.getOrElse(memoryMap.roDataSize.signed)
    if roSize > 0 then
      mappings += ((memoryMap.roDataAddress, UInt(roSize), PageAccess.ReadOnly))

    // RW data region - READ_WRITE
    if memoryMap.rwDataSize.signed > 0 then
      mappings += ((memoryMap.rwDataAddress, memoryMap.rwDataSize, PageAccess.ReadWrite))

    // Stack region - READ_WRITE
    if memoryMap.stackSize.signed > 0 then
      mappings += ((memoryMap.stackAddressLow, memoryMap.stackSize, PageAccess.ReadWrite))

    // Aux data region handling with GP stack
    if memoryMap.auxDataSize.signed > 0 then
      val gpStackLow = PvmConstants.GpStackLow
      val gpStackBase = PvmConstants.GpStackBase
      val gpStackSize = PvmConstants.GpStackSize

      val auxStart = memoryMap.auxDataAddress
      val auxEnd = auxStart + memoryMap.auxDataSize

      // Check if GP stack region overlaps with aux data
      if gpStackLow.toLong >= auxStart.toLong && gpStackBase.toLong <= auxEnd.toLong then
        // Mark before GP stack as READ_ONLY
        if gpStackLow.toLong > auxStart.toLong then
          mappings += ((auxStart, UInt((gpStackLow.toLong - auxStart.toLong).toInt), PageAccess.ReadOnly))

        // GP stack region - READ_WRITE
        mappings += ((gpStackLow, gpStackSize, PageAccess.ReadWrite))

        // Mark after GP stack as READ_ONLY
        if gpStackBase.toLong < auxEnd.toLong then
          mappings += ((gpStackBase, UInt((auxEnd.toLong - gpStackBase.toLong).toInt), PageAccess.ReadOnly))
      else
        // No GP stack overlap - mark entire aux as READ_ONLY
        mappings += ((auxStart, memoryMap.auxDataSize, PageAccess.ReadOnly))

    PageMap.create(mappings.toList, memoryMap.pageSize)

  /**
   * Aligns size to the next page boundary.
   */
  def alignToNextPageSize(pageSize: Int, size: Int): Int =
    ((size + pageSize - 1) / pageSize) * pageSize
